docs/developer/best-practices/navigation.asciidoc

```diff
@@ -0,0 +1,203 @@
+to make sure there is no page reload when navigating from one place in {kib} to another. 
+
+**For example**, navigation using native browser apis:  
+
+[source,js]
+----
+window.location.href = core.http.basePath.prepend(`/dashboard/my-dashboard`); // (try to avoid this)
+----
+
+would cause a full page reload.
+
+To navigate between different {kib} apps without a page reload there are apis in `core`:
+
+* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetoapp.md[core.application.navigateToApp]
+* {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetourl.md[core.application.navigateToUrl]
+
+*Rendering a link to a different {kib} app on its own would also cause a full page reload:*
+
+[source,typescript jsx]
+----
+const myLink = () => <a href={core.http.basePath.prepend(`/dashboard/my-dashboard`)}> Go to Dashboard </a>; 
+----
+
+A workaround could be to handle a click, prevent browser navigation and use `core.application.navigateToApp` api:
+
+[source,typescript jsx]
+----
+const MySPALink = () => 
+    <a 
+        href={core.http.basePath.prepend(`/dashboard/my-dashboard`)} 
+        onClick={(e) => {
+            e.preventDefault();
+            core.application.navigateToApp('dashboard', {path: '/my-dashboard'}); 
+        }}> 
+        Go to Dashboard 
+    </a>; 
+----
+
+
+As it would be too much boilerplate to do this for each {kib} link in your app, there is a handy HOC that helps with it: 
+{kib-repo}tree/{branch}/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L49[RedirectAppLinks].
+
+[source,typescript jsx]
+----
+const MyApp = () => 
+  <RedirectAppLinks application={core.application}>
+    {/*...*/}
+    {/* navigations using this link will happen in SPA friendly way */}
+        <a href={core.http.basePath.prepend(`/dashboard/my-dashboard`)}>Go to Dashboard</a>
+    {/*...*/}
+  </RedirectAppLinks>
+----
+
+
+[[deep-linking]]
+=== Deep-linking into {kib} apps
+
+**Consider a {kib} app URL a part of app's plugin contract.** +
+Try to avoid hardcoding other app's URL in your app's code:
+
+[source,typescript jsx]
+----
+const urlToDashboarWithComplexState = core.http.basePath.prepend(`/dashboards#/view/0289f5d0-e61c-11ea-8c31-8bd3114ed023?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-15m,to:now))&_a=(description:'',filters:!(),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!f,title:'URL%20Drilldown%20Demo',viewMode:view)`)
+----
+
+Instead, each app should expose {kib-repo}tree/{branch}/src/plugins/share/public/url_generators/README.md[a URL generator].
+Other apps should use those URL generators for deep-linking.
+To get a better idea, take a look at {kib-repo}tree/{branch}/src/plugins/dashboard/public/url_generator.ts#L38[dashboard's app URL generator].
```

Comment:
Maybe better to show Discover URL generator as example here as it is newer and cleaner.

- https://github.com/elastic/kibana/blob/master/src/plugins/discover/public/url_generator.ts